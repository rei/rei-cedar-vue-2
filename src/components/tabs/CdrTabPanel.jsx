import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import style from './styles/CdrTabPanel.scss';

export default {
  name: 'CdrTabPanel',
  mixins: [modifier],
  props: {
    /**
     * Required string value that shows up on tab header
     */
    name: {
      type: String,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    ariaLabelledby: {
      type: String,
      required: true,
    },
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      active: false,
      hidden: true,
      offsetX: 0,
      animationDirection: null,
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tab-panel';
    },
    animationClass() {
      return this.animationDirection ? style[`cdr-tab-panel-${this.animationDirection}`] : null;
    },
  },
  methods: {
    setActive(state) {
      if (state) this.hidden = false;
      this.active = state;
      this.$emit('tab-change', state, this.id);
      this.$emit('tabChange', state, this.id);
    },
    setAnimationDirection(direction) {
      this.animationDirection = direction;
    },
    setOffsetX(x) {
      this.offsetX = x;
    },
    handleUpArrowNav() {
      this.$parent.setFocusToActiveTabHeader();
    },
    animationEnd(event) {
      if (event.animationName.split('-')[0] === 'exit') {
        this.hidden = true;
        this.animationDirection = null;
      }
    },
  },
  render() {
    return (
      <div
        aria-hidden={!this.active}
        aria-labelledby={this.ariaLabelledby}
        class={clsx(this.style[this.baseClass], this.modifierClass, this.animationClass)}
        hidden={this.hidden}
        id={this.id}
        ref="cdrTabPanelContainer"
        tabindex="0"
        role="tabpanel"
        vOn:keydown_up_prevent={this.handleUpArrowNav}
        vOn:animationend={this.animationEnd}
        key={this.name}
      >
        {this.$slots.default}
      </div>
    );
  },
};
