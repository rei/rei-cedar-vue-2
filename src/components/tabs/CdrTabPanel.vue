<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <transition
    name="fly"
    v-on="animationHooks">
    <div
      v-show="active"
      :aria-hidden="!active"
      :class="[modifierClass]"
      tabindex="1"
      role="tabpanel"
      :key="name">
      <slot/>
    </div>
  </transition>
</template>

<script>
import modifier from 'mixinsdir/modifier';

export default {
  name: 'CdrTabPanel',
  mixins: [modifier],
  props: {
    name: String,
    id: String,
  },
  data() {
    return {
      active: false,
      offsetX: 0,
      tabId: this.id || this.name,
      animationDirection: 'default',
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tab-panel';
    },
    animationHooks() {
      return {
        beforeEnter: this.setEnterStart,
        afterEnter: this.setEnterEnd,
        beforeLeave: this.setLeaveStart,
        afterLeave: this.setLeaveEnd,
        ...this.$listeners,
      };
    },
  },
  methods: {
    setActive(state) {
      this.active = state;
      this.$emit('tabChange', state, this.tabId);
    },
    setAnimationDirection(direction) {
      this.animationDirection = direction;
    },
    setOffsetX(x) {
      this.offsetX = x;
    },
    setEnterStart(element) {
      const el = element;
      el.style.animationDirection = 'reverse';
      el.style.animationTimingFunction = 'cubic-bezier(0.4, 0, 0.68, .06)';
      el.classList.add(this.animationDirection);
    },
    setEnterEnd(element) {
      const el = element;
      el.style.animationDirection = '';
      el.classList.remove(this.animationDirection);
    },
    setLeaveStart(element) {
      const el = element;
      el.classList.add(this.animationDirection);
      el.style.animationTimingFunction = 'cubic-bezier(0.32, 0.94, 0.6, 1)';
    },
    setLeaveEnd(element) {
      const el = element;
      el.classList.remove(this.animationDirection);
    },
  },
};
</script>

<style>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrTabPanel.pcss';
</style>
