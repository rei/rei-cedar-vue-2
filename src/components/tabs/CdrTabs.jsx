import debounce from 'lodash-es/debounce';
import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import size from '../../mixins/size';
import style from './styles/CdrTabs.scss';

export default {
  name: 'CdrTabs',
  mixins: [modifier, size],
  props: {
    height: {
      type: String,
      default: '240px',
    },
    activeTab: {
      type: Number,
      default: 0,
    },
  },
  data() {
    return {
      tabs: [],
      underlineOffsetX: 0,
      underlineWidth: 0,
      underlineScrollX: 0,
      activeTabIndex: 0,
      headerWidth: 0,
      headerOverflow: false,
      overflowLeft: false,
      overflowRight: false,
      animationInProgress: false,
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tabs';
    },
    underlineStyle() {
      return {
        transform: `translateX(${this.underlineOffsetX}px)`,
        width: `${this.underlineWidth}px`,
      };
    },
  },
  mounted() {
    this.tabs = (this.$slots.default || [])
      .map(vnode => vnode.componentInstance)
      .filter(tab => tab); // get vue component children in the slot

    this.activeTabIndex = this.getNextTab(this.activeTab);

    if (this.tabs[this.activeTabIndex] && this.tabs[this.activeTabIndex].setActive) {
      this.tabs[this.activeTabIndex].setActive(true);
    }

    this.$nextTick(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      setTimeout(() => {
        this.updateUnderline();
      }, 100);
    });
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      this.updateUnderline();
    }, 500));
    // Check for header overflow on widow resize for gradient behavior.
    this.$refs.cdrTabsHeader.parentElement.addEventListener('scroll', debounce(() => {
      this.calculateOverflow();
      this.updateUnderline();
    }, 250));
  },
  methods: {
    getNextTab(startIndex) {
      for (let i = startIndex; i < this.tabs.length; i += 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== 0) {
        for (let k = 0; k < startIndex; k += 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    getPreviousTab(startIndex) {
      for (let i = startIndex; i > -1; i -= 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== this.tabs.length - 1) {
        for (let k = this.tabs.length - 1; k > startIndex; k -= 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    handleClick: debounce(function handleClickCallback(tabClicked) {
      const newIndex = this.tabs.findIndex(tab => tabClicked.name === tab.name);
      this.changeTab(newIndex);
    }, 500, { leading: true, trailing: false }),
    changeTab(newIndex) {
      const oldIndex = this.activeTabIndex;

      this.hideScrollBar();
      if (newIndex > oldIndex) {
        this.tabs[oldIndex].setAnimationDirection('exit-left');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-right');
        }, 200);
      } else {
        this.tabs[oldIndex].setAnimationDirection('exit-right');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-left');
        }, 200);
      }
      this.activeTabIndex = newIndex;
      this.updateUnderline();
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    rightArrowNav: debounce(function handleRightArrow() {
      const nextTab = this.getNextTab(this.activeTabIndex + 1);
      if (nextTab !== -1) {
        this.changeTab(nextTab);
      }
    }, 300, { leading: true, trailing: false }),
    leftArrowNav: debounce(function handleLeftArrow() {
      const previousTab = this.getPreviousTab(this.activeTabIndex - 1);
      if (previousTab !== -1) {
        this.changeTab(previousTab);
      }
    }, 300, { leading: true, trailing: false }),
    calculateOverflow() {
      let containerWidth = 0;
      if (this.$refs.cdrTabsContainer) {
        containerWidth = this.$refs.cdrTabsContainer.offsetWidth;
      }
      this.headerOverflow = this.headerWidth > containerWidth;
      if (this.headerOverflow) {
        // Get Scroll Position
        const scrollX = this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.overflowLeft = scrollX > 1;
        this.overflowRight = (scrollX + 1) < (this.headerWidth - containerWidth);
      } else {
        this.overflowLeft = false;
        this.overflowRight = false;
      }
    },
    updateUnderline() {
      const elements = Array.from(this.$refs.cdrTabsHeader.children);
      if (elements.length > 0) {
        const activeTab = elements[this.activeTabIndex];
        this.underlineOffsetX = activeTab.offsetLeft
          - this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.underlineWidth = activeTab.firstChild.offsetWidth;
      }
    },
    handleDownArrowNav() {
      if (!this.animationInProgress) {
        this.$el.lastElementChild.children[this.activeTabIndex].focus();
      }
    },
    setFocusToActiveTabHeader() {
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    getHeaderWidth() {
      let headerElements = [];
      if (this.$refs.cdrTabsHeader) {
        headerElements = Array.from(this.$refs.cdrTabsHeader.children);
      }
      let totalWidth = 0;
      headerElements.forEach((element) => {
        totalWidth += element.offsetWidth || 0;
      });
      return totalWidth;
    },
    hideScrollBar() {
      const styleRef = this.$refs.cdrTabsContainer.style;
      window.addEventListener('transitionend', () => {
        styleRef.setProperty('overflow-x', 'unset');
      }, { once: true });
      styleRef.setProperty('overflow-x', 'hidden');
    },
    getTabEl(tab) {
      return tab.disabled ? (
        <span
          class={clsx(
            this.style['cdr-tabs__header-item-label'],
            this.style['cdr-tabs__header-item-label--disabled'],
          )}
          aria-disabled="true"
          aria-selected="false"
        >
          {tab.name}
        </span>
      ) : (
        <a
          tabIndex={tab.active ? 0 : -1}
          vOn:click_prevent={e => this.handleClick(tab, e)}
          href={`#${tab.id}`}
          class={this.style['cdr-tabs__header-item-label']}
        >
          { tab.name }
        </a>
      );
    },
  },
  render() {
    return (
      <div
        class={clsx(this.style[this.baseClass], this.modifierClass, this.sizeClass)}
        ref="cdrTabsContainer"
        style={{ height: this.height }}
      >
        <div
          class={clsx(
            this.overflowLeft ? this.style['cdr-tabs__header-gradient-left'] : '',
            this.overflowRight ? this.style['cdr-tabs__header-gradient-right'] : '',
            this.style['cdr-tabs__gradient-container'],
          )}
          vOn:keyup_right={this.rightArrowNav}
          vOn:keyup_left={this.leftArrowNav}
          vOn:keydown_down_prevent={this.handleDownArrowNav}
        >
          <nav
            class={clsx(
              this.overflowLeft ? this.style['cdr-tabs__header-gradient-left'] : '',
              this.overflowRight ? this.style['cdr-tabs__header-gradient-right'] : '',
              this.style['cdr-tabs__header-container'],
            )}
          >
            <ol
              class={this.style['cdr-tabs__header']}
              role="tablist"
              ref="cdrTabsHeader"
            >
              {this.tabs.map(tab => (
                  <li
                    role="tab"
                    aria-selected={tab.active}
                    aria-disabled="false"
                    aria-controls={tab.id}
                    id={tab.ariaLabelledby}
                    key={tab.id}
                    class={clsx(
                      tab.active ? this.style['cdr-tabs__header-item-active'] : '',
                      this.style['cdr-tabs__header-item'],
                    )}
                  >
                    {this.getTabEl(tab)}
                  </li>
              ))}
            </ol>
          </nav>

          <div
            class={this.style['cdr-tabs__underline']}
            style={this.underlineStyle}
          />
        </div>
        <div
          class={this.style['cdr-tabs__content-container']}
          ref="slotWrapper"
        >
          {this.$slots.default}
        </div>
      </div>
    );
  },
};
