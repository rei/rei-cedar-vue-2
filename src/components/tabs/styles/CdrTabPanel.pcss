/* ==========================================================================
  # CdrTabPanel

  TOC:

    :Base - CdrTabPanel

========================================================================== */

.cdr-tab-panel {
  position: absolute;
  width: 100%;
  height: 100%;
}

@keyframes flyLeft {
  from {
    transform: translateX(0);
  }

  to {
    transform: translateX(-100vw);
  }
}

@keyframes flyRight {
  from {
    transform: translateX(0);
  }

  to {
    transform: translateX(100vw);
  }
}

.flyLeft {
  animation-duration: 0.5s;

  /* animation-timing-function: cubic-bezier(0.32, 0.94, 0.6, 1); */
  animation-name: flyLeft;
}

.flyRight {
  animation-duration: 0.5s;

  /* animation-timing-function: cubic-bezier(0.32, 0.94, 0.6, 1); */
  animation-name: flyRight;
}
