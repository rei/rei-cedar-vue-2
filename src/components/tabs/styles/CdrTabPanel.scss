@import '../../../css/settings/index.scss';

.cdr-tab-panel {
  position: absolute;
  width: 100%;
  height: 100%;
  padding-top: $cdr-space-eighth-x;
}

@keyframes :global(exit-left) {
  from {
    transform: translateX(0);
    opacity: 1;
  }
  
  to {
    transform: translateX(-10px);
    opacity: 0;
  }
}

@keyframes :global(exit-right) {
  from {
    transform: translateX(0);
    opacity: 1;
  }

  to {
    transform: translateX(10px);
    opacity: 0;
  }
}

@keyframes :global(enter-left) {
  from {
    transform: translateX(-10px);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@keyframes :global(enter-right) {
  from {
    transform: translateX(10px);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@mixin cdr-tab-panel-animation {
  animation-duration: $cdr-duration-2-x;
  animation-timing-function: $cdr-timing-function-ease-out;
}

.cdr-tab-panel-enter-left {
  :global {
    animation-name: enter-left;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-enter-right {
  :global {
    animation-name: enter-right;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-exit-left {
  :global {
    animation-name: exit-left;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-exit-right {
  :global {
    animation-name: exit-right;
  }
  @include cdr-tab-panel-animation;
}

