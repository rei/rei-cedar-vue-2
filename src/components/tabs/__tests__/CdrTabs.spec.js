import { mount } from '@vue/test-utils';
import CdrTabs from 'componentdir/tabs/CdrTabs';
import CdrTabPanel from 'componentdir/tabs/CdrTabPanel';
import Vue from 'vue';

// Tests use nextTick because of the nextTick in mounted hook of tabs

describe('CdrTabs', () => {
  describe('mounted', () => {
    it('mounts tabs', () => {
      const wrapper = mount(CdrTabs);
      expect(wrapper.element).toMatchSnapshot();
    });

    it('mounts with cdr-tab-panel children', (done) => {
      const spyGetNextTab = jest.fn();
      const spyGetHeaderWidth = jest.fn();
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ],
        },
        methods: {
          getNextTab: spyGetNextTab,
          getHeaderWidth: spyGetHeaderWidth,
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
        },
      });

      Vue.nextTick(() => {
        expect(wrapper.vm.tabs.length).toBe(2);
        expect(wrapper.findAll('li').length).toBe(2);
        expect(wrapper.element).toMatchSnapshot();
        expect(spyGetNextTab).toHaveBeenCalled();
        expect(spyGetHeaderWidth).toHaveBeenCalled();
        expect(spyCalculateOverflow).toHaveBeenCalled();
        
        setTimeout(() => {
          expect(spyUpdateUnderline).toHaveBeenCalled();
          done();
        }, 200)
      });
    });
  });

  describe('event listeners', () => {    
    it('handles scroll event', (done) => {
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ],
        },
        methods: {
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        wrapper.vm.$refs.cdrTabsHeader.parentElement.dispatchEvent(new Event('scroll'));

        setTimeout(() => { // for debounce
          expect(wrapper.vm.overflowLeft).toBe(false);
          expect(spyCalculateOverflow).toHaveBeenCalled();
          expect(spyUpdateUnderline).toHaveBeenCalled();
          wrapper.destroy();
          done();
        }, 600);
      });
    });

    it('handles resize event', () => {
      const spyCalculateOverflow = jest.fn();
      const spyUpdateUnderline = jest.fn();
      const spyGetHeaderWidth = jest.fn();

      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ],
        },
        methods: {
          calculateOverflow: spyCalculateOverflow,
          updateUnderline: spyUpdateUnderline,
          getHeaderWidth: spyGetHeaderWidth,
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        setTimeout(() => {
          console.log('resize dispatchEvent');
          window.dispatchEvent(new Event('resize'));

          setTimeout(() => { // for debounce
            expect(spyGetHeaderWidth).toHaveBeenCalledTimes(2);
            expect(spyCalculateOverflow).toHaveBeenCalledTimes(2);
            expect(spyUpdateUnderline).toHaveBeenCalledTimes(2);
            wrapper.destroy();
            done();
          }, 600);
        }, 800);
      });
    });
  });

  it('getNextTab and getPreviousTab', () => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />',
          '<cdr-tab-panel name="tab3" id="tab-panel-3" :disabled="true" aria-labelledby="tab-3"  />',
        ]
      }
    });

    expect(wrapper.vm.getNextTab(0)).toBe(0);
    expect(wrapper.vm.getNextTab(1)).toBe(1);
    expect(wrapper.vm.getNextTab(3)).toBe(0);

    expect(wrapper.vm.getPreviousTab(0)).toBe(0);
    expect(wrapper.vm.getPreviousTab(1)).toBe(1);
    expect(wrapper.vm.getPreviousTab(-1)).toBe(1);
  });

  describe('handleClick', () => {
    it('left to right', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ],
        },
      });
      
      Vue.nextTick(() => {
        wrapper.findAll('a').at(1).trigger('click');
        expect(wrapper.vm.activeTabIndex).toBe(1);
        done();
      });
    });

    it('right to left', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-3" aria-labelledby="tab-3" />'
          ],
        },
        propsData: {
          activeTab: 1,
        },
      });
      
      Vue.nextTick(() => {
        expect(wrapper.vm.activeTabIndex).toBe(1);
        wrapper.vm.changeTab(0);
        Vue.nextTick(() => {
          setTimeout(() => {
            expect(wrapper.vm.activeTabIndex).toBe(0)
            done();
          }, 300);
        });
      });
    });
  });

  it('handles right arrow key', (done) => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      }
    });
    
    Vue.nextTick(() => {
      // Trigger right arrow keyup event
      wrapper.findAll('div').at(1).trigger('keyup.right');
      expect(wrapper.vm.activeTabIndex).toBe(1);
      done();
    });
  });

  it('handles left arrow key', (done) => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      }
    });
    
    Vue.nextTick(() => {
      wrapper.setData({ activeTabIndex: 1 });
      // Trigger left arrow keyup event
      wrapper.findAll('div').at(1).trigger('keyup.left');
      expect(wrapper.vm.activeTabIndex).toBe(0);
      done();
    });
  });

  describe('overflow classes', () => {
    it('adds gradient-left class', (done) => {
      const spyUpdateUnderline = jest.fn();
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ],
        },
        methods: {
          updateUnderline: spyUpdateUnderline,
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        wrapper.setData({ overflowLeft: true });
        expect(wrapper.find('.cdr-tabs__header-gradient-left').exists()).toBe(true);
        done();
      });
    });

    it('adds gradient-right class', (done) => {
      const wrapper = mount(CdrTabs, {
        stubs: {
          'cdr-tab-panel': CdrTabPanel,
        },
        slots: {
          default: [
            '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
            '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
          ]
        },
        attachToDocument: true,
      });

      Vue.nextTick(() => {
        wrapper.setData({ overflowRight: true });
        expect(wrapper.find('.cdr-tabs__header-gradient-right').exists()).toBe(true);
        done();
      });
    });
  });

  it('accessibility', (done) => {
    const spyUpdateUnderline = jest.fn();
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1"  aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />',
          '<cdr-tab-panel name="tab3" :disabled="true" id="tab-panel-3" aria-labelledby="tab-3" />']
      },
      methods: {
        updateUnderline: spyUpdateUnderline,
      },
    });

    Vue.nextTick(() => {
      const tab1 = wrapper.find('#tab-1');

      expect(tab1.attributes()['aria-selected']).toBe('true');
      expect(tab1.attributes()['role']).toBe('tab');
      expect(tab1.attributes()['aria-disabled']).toBe('false');

      // tablist role
      expect(wrapper.find({ref: 'cdrTabsHeader'}).attributes()['role']).toBe('tablist');
      
      done();
    });

  });

  it('handleDownArrowNav', () => {
    const spyUpdateUnderline = jest.fn();
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      },
      methods: {
        updateUnderline: spyUpdateUnderline,
      },
      attachToDocument: true,
    });

    Vue.nextTick(() => {
      wrapper.vm.handleDownArrowNav();
      expect(wrapper.vm.$el.lastElementChild.children[wrapper.vm.activeTabIndex]).toBe(document.activeElement);
    });
  });

  it('setFocusToActiveTabHeader', () => {
    const spyUpdateUnderline = jest.fn();
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      },
      methods: {
        updateUnderline: spyUpdateUnderline,
      },
      attachToDocument: true,
    });

    Vue.nextTick(() => {
      wrapper.vm.setFocusToActiveTabHeader();
      expect(wrapper.vm.$refs.cdrTabsHeader.children[wrapper.vm.activeTabIndex].children[0]).toBe(document.activeElement);
    });
  });

  it('scrollbar is hidden properly', (done) => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      },
      attachToDocument: true,
    });
    
    wrapper.setData({ widthInitialized: true});
    wrapper.setData({ underlineWidth: -1});
    wrapper.vm.hideScrollBar();
    expect(wrapper.vm.$refs.cdrTabsContainer.style.getPropertyValue('overflow-x')).toBe('hidden');
    window.dispatchEvent(new Event('transitionend'));
    Vue.nextTick(() => {
      expect(wrapper.vm.$refs.cdrTabsContainer.style.getPropertyValue('overflow-x')).toBe('unset');
      wrapper.destroy();
      done();
    });
  });

  it('calculateOverflow', () => {
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: [
          '<cdr-tab-panel name="tab1" id="tab-panel-1" aria-labelledby="tab-1" />',
          '<cdr-tab-panel name="tab2" id="tab-panel-2" aria-labelledby="tab-2" />'
        ],
      }
    });
    wrapper.setData({ headerWidth: 2000 });
    wrapper.vm.calculateOverflow();
    expect(wrapper.vm.headerOverflow).toBe(true);
  });
});
