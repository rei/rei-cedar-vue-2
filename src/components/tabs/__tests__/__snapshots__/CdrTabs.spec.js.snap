// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrTabs mounted mounts tabs 1`] = `
<div
  class="cdr-tabs"
  style="height: 240px;"
>
  <div
    class="cdr-tabs__gradient-container"
  >
    <nav
      class="cdr-tabs__header-container"
    >
      <ol
        class="cdr-tabs__header"
        role="tablist"
      />
    </nav>
    <div
      class="cdr-tabs__underline"
      style="transform: translateX(0px); width: 0px;"
    />
  </div>
  <div
    class="cdr-tabs__content-container"
  />
</div>
`;

exports[`CdrTabs mounted mounts with cdr-tab-panel children 1`] = `
<div
  class="cdr-tabs"
  style="height: 240px;"
>
  <div
    class="cdr-tabs__gradient-container"
  >
    <nav
      class="cdr-tabs__header-container"
    >
      <ol
        class="cdr-tabs__header"
        role="tablist"
      >
        <li
          aria-controls="tab-panel-1"
          aria-disabled="false"
          class="cdr-tabs__header-item"
          id="tab-1"
          role="tab"
        >
          <a
            class="cdr-tabs__header-item-label"
            href="#tab-panel-1"
            tabindex="-1"
          >
            tab1
          </a>
        </li>
        <li
          aria-controls="tab-panel-2"
          aria-disabled="false"
          class="cdr-tabs__header-item"
          id="tab-2"
          role="tab"
        >
          <a
            class="cdr-tabs__header-item-label"
            href="#tab-panel-2"
            tabindex="-1"
          >
            tab2
          </a>
        </li>
      </ol>
    </nav>
    <div
      class="cdr-tabs__underline"
      style="transform: translateX(0px); width: 0px;"
    />
  </div>
  <div
    class="cdr-tabs__content-container"
  >
    <div
      aria-hidden="true"
      aria-labelledby="tab-1"
      class="cdr-tab-panel"
      hidden="hidden"
      id="tab-panel-1"
      role="tabpanel"
      tabindex="0"
    />
    <div
      aria-hidden="true"
      aria-labelledby="tab-2"
      class="cdr-tab-panel"
      hidden="hidden"
      id="tab-panel-2"
      role="tabpanel"
      tabindex="0"
    />
  </div>
</div>
`;
